panic: value of type *strings.Reader is not assignable to type io.ReadCloser

The following program `sample.go` triggers a panic:

```go
package main

import (
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"strings"
)

type readAutoCloser struct {
	r io.ReadCloser
}

func (a readAutoCloser) Read(b []byte) (n int, err error) {
	if a.r == nil {
		return 0, io.EOF
	}
	n, err = a.r.Read(b)
	if err == io.EOF {
		a.Close()
	}
	return n, err
}

func (a readAutoCloser) Close() error {
	if a.r == nil {
		return nil
	}
	return a.r.(io.Closer).Close()
}

type pipe struct {
	Reader readAutoCloser
}

func newReadAutoCloser(r io.Reader) readAutoCloser {
	if _, ok := r.(io.Closer); !ok {
		return readAutoCloser{ioutil.NopCloser(r)}
	}
	return readAutoCloser{r.(io.ReadCloser)}
}

func main() {
	p := &pipe{}
	p.Reader = newReadAutoCloser(strings.NewReader("test"))
	b, err := ioutil.ReadAll(p.Reader)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println(b)
}
```

Expected result:
```console
$ go run ./sample.go
[116 101 115 116]
```

Got:
```console
main.go:38:17: panic
main.go:45:8: panic
reflect.Set: value of type *strings.Reader is not assignable to type io.ReadCloser
goroutine 1 [running]:
runtime/debug.Stack(0x1, 0xc0002fe200, 0x40)
        /usr/local/Cellar/go/1.14/libexec/src/runtime/debug/stack.go:24 +0x9d
github.com/containous/yaegi/interp.(*Interpreter).Eval.func1(0xc00012de10)
        /Users/john/go/pkg/mod/github.com/containous/yaegi@v0.8.0/interp/interp.go:320 +0xbc
panic(0x188f440, 0xc00006fe60)
        /usr/local/Cellar/go/1.14/libexec/src/runtime/panic.go:967 +0x15d
github.com/containous/yaegi/interp.runCfg.func1(0xc0002ee0a0, 0xc00030cc00)
        /Users/john/go/pkg/mod/github.com/containous/yaegi@v0.8.0/interp/run.go:113 +0x20c
panic(0x188f440, 0xc00006fe60)
        /usr/local/Cellar/go/1.14/libexec/src/runtime/panic.go:967 +0x15d
github.com/containous/yaegi/interp.runCfg.func1(0xc0002ee140, 0xc00030ad00)
        /Users/john/go/pkg/mod/github.com/containous/yaegi@v0.8.0/interp/run.go:113 +0x20c
panic(0x188f440, 0xc00006fe60)
        /usr/local/Cellar/go/1.14/libexec/src/runtime/panic.go:967 +0x15d
reflect.Value.assignTo(0x196d880, 0xc00000ff40, 0x16, 0x19c1190, 0xb, 0x18ec220, 0xc00006fe40, 0x19a8c40, 0x196d880, 0xc00000ff40)
        /usr/local/Cellar/go/1.14/libexec/src/reflect/value.go:2403 +0x426
reflect.Value.Set(0x18ec220, 0xc00006fe40, 0x194, 0x196d880, 0xc00000ff40, 0x16)
        /usr/local/Cellar/go/1.14/libexec/src/reflect/value.go:1532 +0xbd
github.com/containous/yaegi/interp.typeAssert.func1(0xc0002ee140, 0xc00000fc80)
        /Users/john/go/pkg/mod/github.com/containous/yaegi@v0.8.0/interp/run.go:162 +0xba
github.com/containous/yaegi/interp.runCfg(0xc00030ad00, 0xc0002ee140)
        /Users/john/go/pkg/mod/github.com/containous/yaegi@v0.8.0/interp/run.go:119 +0x6d
github.com/containous/yaegi/interp.call.func5(0xc0002ee0a0, 0xc0002edea0)
        /Users/john/go/pkg/mod/github.com/containous/yaegi@v0.8.0/interp/run.go:787 +0xb18
github.com/containous/yaegi/interp.runCfg(0xc00030cc00, 0xc0002ee0a0)
        /Users/john/go/pkg/mod/github.com/containous/yaegi@v0.8.0/interp/run.go:119 +0x6d
github.com/containous/yaegi/interp.(*Interpreter).run(0xc000001200, 0xc00030c300, 0xc0002ee000)
        /Users/john/go/pkg/mod/github.com/containous/yaegi@v0.8.0/interp/run.go:97 +0x239
github.com/containous/yaegi/interp.(*Interpreter).Eval(0xc000001200, 0xc0002f6000, 0x322, 0x0, 0x0, 0x0, 0x0, 0x0)
        /Users/john/go/pkg/mod/github.com/containous/yaegi@v0.8.0/interp/interp.go:388 +0x41f
main.main()
        /Users/john/go/pkg/mod/github.com/containous/yaegi@v0.8.0/cmd/yaegi/yaegi.go:94 +0x548
```

--- a/_test/issue-558.go
+++ b/_test/issue-558.go
@@ -1,0 +1,55 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"strings"
+)
+
+type readAutoCloser struct {
+	r io.ReadCloser
+}
+
+func (a readAutoCloser) Read(b []byte) (n int, err error) {
+	if a.r == nil {
+		return 0, io.EOF
+	}
+	n, err = a.r.Read(b)
+	if err == io.EOF {
+		a.Close()
+	}
+	return n, err
+}
+
+func (a readAutoCloser) Close() error {
+	if a.r == nil {
+		return nil
+	}
+	return a.r.(io.Closer).Close()
+}
+
+type pipe struct {
+	Reader readAutoCloser
+}
+
+func newReadAutoCloser(r io.Reader) readAutoCloser {
+	if _, ok := r.(io.Closer); !ok {
+		return readAutoCloser{ioutil.NopCloser(r)}
+	}
+	return readAutoCloser{r.(io.ReadCloser)}
+}
+
+func main() {
+	p := &pipe{}
+	p.Reader = newReadAutoCloser(strings.NewReader("test"))
+	b, err := ioutil.ReadAll(p.Reader)
+	if err != nil {
+		log.Fatal(err)
+	}
+	fmt.Println(string(b))
+}
+
+// Output:
+// test

--- a/_test/type20.go
+++ b/_test/type20.go
@@ -1,0 +1,18 @@
+package main
+
+import (
+	"io"
+	"strings"
+)
+
+func isCloser(r io.Reader) bool {
+	_, ok := r.(io.Closer)
+	return ok
+}
+
+func main() {
+	println(isCloser(strings.NewReader("test")))
+}
+
+// Output:
+// false

--- a/interp/run.go
+++ b/interp/run.go
@@ -121,30 +121,33 @@ func runCfg(n *node, f *frame) {
 }
 
 func typeAssertStatus(n *node) {
-	value := genValue(n.child[0])      // input value
+	c0, c1 := n.child[0], n.child[1]
+	value := genValue(c0)              // input value
 	value1 := genValue(n.anc.child[1]) // returned status
+	typ := c1.typ.rtype                // type to assert
 	next := getExec(n.tnext)
 
 	switch {
-	case n.child[0].typ.cat == valueT:
+	case c0.typ.cat == valueT:
 		n.exec = func(f *frame) bltn {
-			if !value(f).IsValid() || value(f).IsNil() {
+			v := value(f)
+			if !v.IsValid() || v.IsNil() {
 				value1(f).SetBool(false)
 			}
-			value1(f).SetBool(true)
+			value1(f).SetBool(v.Type().Implements(typ))
 			return next
 		}
-	case n.child[1].typ.cat == interfaceT:
+	case c1.typ.cat == interfaceT:
 		n.exec = func(f *frame) bltn {
 			_, ok := value(f).Interface().(valueInterface)
-			//value0(f).Set(reflect.ValueOf(valueInterface{v.node, v.value}))
+			// TODO: verify that value(f) implements asserted type.
 			value1(f).SetBool(ok)
 			return next
 		}
 	default:
 		n.exec = func(f *frame) bltn {
 			_, ok := value(f).Interface().(valueInterface)
-			//value0(f).Set(v.value)
+			// TODO: verify that value(f) implements asserted type.
 			value1(f).SetBool(ok)
 			return next
 		}
@@ -152,25 +155,29 @@ func typeAssertStatus(n *node) {
 }
 
 func typeAssert(n *node) {
-	value := genValue(n.child[0]) // input value
-	dest := genValue(n)           // returned result
+	c0, c1 := n.child[0], n.child[1]
+	value := genValue(c0) // input value
+	dest := genValue(n)   // returned result
 	next := getExec(n.tnext)
 
 	switch {
-	case n.child[0].typ.cat == valueT:
+	case c0.typ.cat == valueT:
 		n.exec = func(f *frame) bltn {
-			dest(f).Set(value(f).Elem())
+			v := value(f)
+			dest(f).Set(v.Elem())
 			return next
 		}
-	case n.child[1].typ.cat == interfaceT:
+	case c1.typ.cat == interfaceT:
 		n.exec = func(f *frame) bltn {
 			v := value(f).Interface().(valueInterface)
+			// TODO: verify that value(f) implements asserted type.
 			dest(f).Set(reflect.ValueOf(valueInterface{v.node, v.value}))
 			return next
 		}
 	default:
 		n.exec = func(f *frame) bltn {
 			v := value(f).Interface().(valueInterface)
+			// TODO: verify that value(f) implements asserted type.
 			dest(f).Set(v.value)
 			return next
 		}
@@ -195,6 +202,7 @@ func typeAssert2(n *node) {
 	case n.child[1].typ.cat == interfaceT:
 		n.exec = func(f *frame) bltn {
 			v, ok := value(f).Interface().(valueInterface)
+			// TODO: verify that value(f) implements asserted type.
 			value0(f).Set(reflect.ValueOf(valueInterface{v.node, v.value}))
 			value1(f).SetBool(ok)
 			return next
@@ -202,6 +210,7 @@ func typeAssert2(n *node) {
 	default:
 		n.exec = func(f *frame) bltn {
 			v, ok := value(f).Interface().(valueInterface)
+			// TODO: verify that value(f) implements asserted type.
 			value0(f).Set(v.value)
 			value1(f).SetBool(ok)
 			return next

